pbxd/pbx/definity.py

```diff
@@ -0,0 +1,421 @@
+            if index == 0:
+                self.logger.error('Timeout during disconnect:\n{}'.format(self.session.before))
+            elif index == 1:
+                self.logger.error('Connection failed with EOF during disconnect:\n{}'.format(self.session.before))
+            elif index == 2:
+                self.session.sendline('y')
+            self.session.close()
+
+        self.session = None
+        self.connected_termtype = None
+        self.logger.info('Connection closed')
+
+    def reconnect(self):
+        self.logger.warning('Reconnecting...')
+        self.disconnect()
+        self.connect()
+
+    def _select_termtype(self, termtype):
+        """
+        Switch between the ossi4 and vt220 termtypes.
+        """
+        if not self.session.isalive():
+            self.logger.error('dead session: {}'.format(self.session.before))
+            self.reconnect()
+
+        if termtype == self.connected_termtype:
+            return
+
+        if self.connected_termtype == 'vt220':
+            self.session.sendline('newterm')
+        elif self.connected_termtype == 'ossi4':
+            self.session.sendline('c newterm')
+            self.session.sendline('t')
+
+        # Terminal Type (513, 715, 4410, 4425, VT220, NTT, W2KTT, SUNT): [513]
+        index = self.session.expect(
+            [
+                pexpect.TIMEOUT,
+                pexpect.EOF,
+                r'Terminal Type \(.+\): \[.+\]',
+            ]
+        )
+        if index == 0:  # TIMEOUT
+            self.logger.error('Timeout on termtype:\n{}'.format(self.session.before))
+            raise Exception('Timeout on termtype:\n{}'.format(self.session.before))
+        elif index == 1:  # EOF
+            self.logger.error('Connection failed with EOF at termtype:\n{}'.format(self.session.before))
+            error_msg = self.session.before.decode('utf-8').strip().split('\n')[0]  # first output line
+            raise Exception('Connection failed with EOF at termtype: {}'.format(error_msg))
+        elif index == 2:  # Terminal Type
+            self.logger.debug('selecting termtype {} from {}'.format(termtype, self.session.after))
+            self.session.sendline(termtype)
+
+        # verify and consume the prompt
+        if termtype == 'vt220':  # consume the vt220 prompt
+            self.session.expect([r'\x1b\[2;1H.*\x1b\[KCommand: '])
+        elif termtype == 'ossi4':  # consume the ossi t prompt
+            self.session.expect([r't[\r\n]+'])
+        if index == 0:  # TIMEOUT
+            self.logger.error('Timeout on command prompt verify:\n{}'.format(self.session.before))
+            raise Exception('Timeout on command prompt verify:\n{}'.format(self.session.before))
+        elif index == 1:  # EOF
+            self.logger.error('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
+            raise Exception('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
```

Comment:
This block doesn't seem like it's doing what you expect; you aren't updating the `index` based on the most recent `expect`. This is still `index` from above. And since we didn't `raise` above if we got this far these exceptions will never fire.

Reply:
Thanks, you're right! That's a dead code path and I will fix that.